src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+    PICOCACHE().maintenance_register_0.set(0x1);
+}
+
+pub fn enable_picocache(enable : bool) {
+    if enable {
+        PICOCACHE().control.set(0x1);
+    } else {
+        PICOCACHE().control.set(0x0);
+    }
+}
+
+pub fn pico_enabled() -> bool {
+    PICOCACHE().status.get() & 0x1 != 0
+} 
+
+// Struct of the FLASHCALW registers. Section 14.10 of the datasheet
+#[repr(C, packed)]
+#[allow(dead_code)]
+struct Registers {
+    control:                          VolatileCell<u32>,
+    command:                          VolatileCell<u32>,
+    status:                           VolatileCell<u32>,
+    parameter:                        VolatileCell<u32>,
+    version:                          VolatileCell<u32>,
+    general_purpose_fuse_register_hi: VolatileCell<u32>,
+    general_purpose_fuse_register_lo: VolatileCell<u32>,
+}
+
+const FLASHCALW_BASE_ADDRS : *mut Registers = 0x400A0000 as *mut Registers;
+
+#[allow(dead_code)]
+enum RegKey {
+    CONTROL,
+    COMMAND,
+    STATUS,
+    PARAMETER,
+    VERSION,
+    GPFRHI,
+    GPFRLO
+}
+
+/// Error codes are used to inform the Client if the command completed successfully
+/// or whether there was an error and what type of error it was.
+pub enum Error {
+    CommandComplete,    /* Command Complete */
+    LockE,              /* Lock Error (i.e. tried writing to locked page) */
+    ProgE,              /* Program Error (i.e. incorrectly issued flash commands */
+    LockProgE,          /* Lock and Program Error */
+    ECC,                /* Error Correcting Code Error */
+}
+
+/// High level commands to issue to the flash. Usually to track the state of 
+/// a command especially if it's multiple FlashCMDs. 
+/// 
+/// For example an erase is: 1) Unlock Page  (UP)
+///                          2) Erase Page   (EP)
+///                          3) Lock Page    (LP)
+/// Store what high level command we're doing allows us to track the state and
+/// continue the steps of the command in handle_interrupt.
+#[derive(Clone, Copy, PartialEq)]
+pub enum Command {
+    Write,
+    Read,
+    Erase,
+    None
+}
+
+
+
+/// There are 18 recognized commands for the flash. These are 'bare-bones' commands
+/// and values that are written to the Flash's command register to inform 
+/// the flash what to do. Table 14-5.
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashCMD {
+    NOP,
+    WP,
+    EP,
+    CPB,
+    LP,
+    UP,
+    EA,
+    WGPB,
+    EGPB,
+    SSB,
+    PGPFB,
+    EAGPF,
+    QPR,
+    WUP,
+    EUP,
+    QPRUP,
+    HSEN,
+    HSDIS,
+}
+
+//The two Flash speeds
+#[derive(Clone, Copy)]
+pub enum Speed {
+    Standard,
+    HighSpeed
+}
+
+/// FlashState is used to track the current state of the flash in high level
+/// command.
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashState {
+    Locking,
+    Unlocking,
+    WritePageBuffer,
+    Writing,
+    Reading,
+    Erasing,
+    Ready,
+    Unconfigured
+}
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
```

Comment:
In general, it's not clear what the use of several of the state fields are. I think some of them could/should be removed, but those you choose to retain, it would be useful to comment what they are used for and what invariant should be maintained.